File: data/src/main/java/org/cryptomator/data/cloud/webdav/WebDavImpl.java
package org.cryptomator.data.cloud.webdav;

import android.content.Context;

import org.cryptomator.data.cloud.webdav.network.ConnectionHandlerHandlerImpl;
import org.cryptomator.data.util.CopyStream;
import org.cryptomator.data.util.TransferredBytesAwareInputStream;
import org.cryptomator.data.util.TransferredBytesAwareOutputStream;
import org.cryptomator.domain.CloudFile;
import org.cryptomator.domain.CloudFolder;
import org.cryptomator.domain.CloudNode;
import org.cryptomator.domain.WebDavCloud;
import org.cryptomator.domain.exception.BackendException;
import org.cryptomator.domain.exception.CloudNodeAlreadyExistsException;
import org.cryptomator.domain.exception.FatalBackendException;
import org.cryptomator.domain.exception.NotFoundException;
import org.cryptomator.domain.exception.ParentFolderDoesNotExistException;
import org.cryptomator.domain.usecases.ProgressAware;
import org.cryptomator.domain.usecases.cloud.DataSource;
import org.cryptomator.domain.usecases.cloud.DownloadState;
import org.cryptomator.domain.usecases.cloud.Progress;
import org.cryptomator.domain.usecases.cloud.UploadState;
import org.cryptomator.util.Optional;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.List;

import okhttp3.HttpUrl;

import static org.cryptomator.domain.usecases.cloud.Progress.progress;

class WebDavImpl {

	private final WebDavCloud cloud;
	private final HttpUrl baseUrl;
	private final RootWebDavFolder root;
	private final ConnectionHandlerHandlerImpl connectionHandler;

	WebDavImpl(WebDavCloud cloud, ConnectionHandlerHandlerImpl connectionHandler) {
		this.cloud = cloud;
		this.baseUrl = HttpUrl.parse(cloud.url());
		this.root = new RootWebDavFolder(cloud);
		this.connectionHandler = connectionHandler;
	}

	public WebDavFolder root() {
		return root;
	}

	public WebDavFolder resolve(String path) {
		if (path.startsWith("/")) {
			path = path.substring(1);
		}
		String[] names = path.split("/");
		WebDavFolder folder = root;
		for (String name : names) {
			folder = folder(folder, name);
		}
		return folder;
	}

	public WebDavFile file(CloudFolder parent, String name) {
		return file(parent, name, Optional.empty());
	}

	public WebDavFile file(CloudFolder parent, String name, Optional<Long> size) {
		return new WebDavFile((WebDavFolder) parent, name, parent.getPath() + '/' + name, size, Optional.empty());
	}

	public WebDavFolder folder(CloudFolder parent, String name) {
		return new WebDavFolder((WebDavFolder) parent, name, parent.getPath() + '/' + name);
	}

	public boolean exists(CloudNode node) throws BackendException {
		try {
			return connectionHandler //
					.get(absoluteUriFrom(node.getPath()), //
							node.getParent()) != null;
		} catch (NotFoundException e) {
			return false;
		}
	}

	public List<CloudNode> list(WebDavFolder folder) throws BackendException {
		return connectionHandler //
				.dirList(absoluteUriFrom(folder.getPath()), //
						folder);
	}

	public WebDavFolder create(WebDavFolder folder) throws BackendException {
		try {
			return createExcludingParents(folder);
		} catch (NotFoundException | ParentFolderDoesNotExistException e) {
			create(folder.getParent());
			return createExcludingParents(folder);
		}
	}

	private WebDavFolder createExcludingParents(WebDavFolder folder) throws BackendException {
		if (folder.getParent() == null) {
			return folder;
		} else {
			return connectionHandler.createFolder( //
					absoluteUriFrom(folder.getPath()), //
					folder);
		}
	}

	public WebDavFolder move(CloudFolder source, CloudFolder target) throws BackendException {
		moveFileOrFolder(source, target);
		return new WebDavFolder( //
				(WebDavFolder) target.getParent() //
				, target.getName() //
				, target.getPath());
	}

	public WebDavFile move(CloudFile source, CloudFile target) throws BackendException {
		moveFileOrFolder(source, target);
		return new WebDavFile( //
				(WebDavFolder) target.getParent() //
				, target.getName() //
				, target.getPath() //
				, source.getSize() //
				, source.getModified());
	}

	private void moveFileOrFolder(CloudNode source, CloudNode target) throws BackendException {
		if (exists(target)) {
			throw new CloudNodeAlreadyExistsException(target.getName());
		}

		connectionHandler //
				.move(absoluteUriFrom(source.getPath()), //
						absoluteUriFrom(target.getPath()));
	}

	public WebDavFile write(final WebDavFile uploadFile, DataSource data, final ProgressAware<UploadState> progressAware, boolean replace, final long size) //
			throws BackendException, IOException {
		if (!replace && exists(uploadFile)) {
			throw new CloudNodeAlreadyExistsException("CloudNode already exists and replace is false");
		}

		progressAware.onProgress(Progress.started(UploadState.upload(uploadFile)));

		try (TransferredBytesAwareDataSource out = new TransferredBytesAwareDataSource(data) {
			@Override
			public void bytesTrasferred(long transferred) {
				progressAware.onProgress( //
						progress(UploadState.upload(uploadFile)) //
								.between(0) //
								.and(size) //
								.withValue(transferred));
			}
		}) {
			connectionHandler //
					.writeFile( //
							absoluteUriFrom(uploadFile.getPath()), out);
		}

		WebDavFile cloudFile = (WebDavFile) connectionHandler //
				.get(absoluteUriFrom(uploadFile.getPath()), //
						uploadFile.getParent());

		if (cloudFile == null) {
			throw new FatalBackendException("Unable to get CloudFile after upload.");
		}

		return cloudFile;
	}

	public void checkAuthenticationAndServerCompatibility(String url) throws BackendException {
		connectionHandler.checkAuthenticationAndServerCompatibility(url);
	}

	public void read(final CloudFile file, OutputStream data, final ProgressAware<DownloadState> progressAware) throws BackendException, IOException {
		progressAware.onProgress(Progress.started(DownloadState.download(file)));

		try (InputStream in = connectionHandler.readFile(absoluteUriFrom(file.getPath())); //
			 TransferredBytesAwareOutputStream out = new TransferredBytesAwareOutputStream(data) {
				 @Override
				 public void bytesTransferred(long transferred) {
					 progressAware.onProgress( //
							 progress(DownloadState.download(file)) //
									 .between(0) //
									 .and(file.getSize().orElse(Long.MAX_VALUE)) //
									 .withValue(transferred));
				 }
			 }) {
			CopyStream.copyStreamToStream(in, out);
		}

		progressAware.onProgress(Progress.completed(DownloadState.download(file)));
	}

	public void delete(CloudNode node) throws BackendException {
		connectionHandler.delete(absoluteUriFrom(node.getPath()));
	}

	private String absoluteUriFrom(String path) {
		path = removeLeadingSlash(path);

		return baseUrl.newBuilder() //
				.addPathSegments(path) //
				.build() //
				.toString();
	}

	private String removeLeadingSlash(String path) {
		return path.length() > 0 && path.charAt(0) == '/' ? path.substring(1) : path;
	}

	public String currentAccount() throws BackendException {
		checkAuthenticationAndServerCompatibility(cloud.url());
		return cloud.url();
	}

	private static abstract class TransferredBytesAwareDataSource implements DataSource {

		private final DataSource data;

		TransferredBytesAwareDataSource(DataSource data) {
			this.data = data;
		}

		@Override
		public Optional<Long> size(Context context) {
			return data.size(context);
		}

		@Override
		public InputStream open(Context context) throws IOException {
			return new TransferredBytesAwareInputStream(data.open(context)) {
				@Override
				public void bytesTransferred(long transferred) {
					TransferredBytesAwareDataSource.this.bytesTrasferred(transferred);
				}
			};
		}

		@Override
		public void close() throws IOException {
			data.close();
		}

		public abstract void bytesTrasferred(long transferred);

		@Override
		public DataSource decorate(DataSource delegate) {
			return delegate;
		}
	}
}

File: data/src/main/java/org/cryptomator/data/cloud/webdav/WebDavImpl.kt
package org.cryptomator.data.cloud.webdav

import android.content.Context
import org.cryptomator.data.cloud.webdav.network.ConnectionHandlerHandlerImpl
import org.cryptomator.data.cloud.webdav.network.ServerNotWebdavCompatibleException
import org.cryptomator.data.util.CopyStream
import org.cryptomator.data.util.TransferredBytesAwareInputStream
import org.cryptomator.data.util.TransferredBytesAwareOutputStream
import org.cryptomator.domain.CloudFile
import org.cryptomator.domain.CloudNode
import org.cryptomator.domain.WebDavCloud
import org.cryptomator.domain.exception.BackendException
import org.cryptomator.domain.exception.CloudNodeAlreadyExistsException
import org.cryptomator.domain.exception.FatalBackendException
import org.cryptomator.domain.exception.NotFoundException
import org.cryptomator.domain.exception.ParentFolderDoesNotExistException
import org.cryptomator.domain.exception.ParentFolderIsNullException
import org.cryptomator.domain.exception.authentication.WebDavNotSupportedException
import org.cryptomator.domain.usecases.ProgressAware
import org.cryptomator.domain.usecases.cloud.DataSource
import org.cryptomator.domain.usecases.cloud.DownloadState
import org.cryptomator.domain.usecases.cloud.Progress
import org.cryptomator.domain.usecases.cloud.UploadState
import java.io.IOException
import java.io.OutputStream
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull

internal class WebDavImpl(private val cloud: WebDavCloud, private val connectionHandler: ConnectionHandlerHandlerImpl, private val context: Context) {

	private val baseUrl: HttpUrl = cloud.url().toHttpUrlOrNull() ?: throw FatalBackendException("Cloud url shouldn't be null")

	private val root: RootWebDavFolder = RootWebDavFolder(cloud)

	fun root(): WebDavFolder {
		return root
	}

	fun resolve(path: String): WebDavFolder {
		val names = path.removePrefix("/").split("/").toTypedArray()
		var folder: WebDavFolder = root
		for (name in names) {
			folder = folder(folder, name)
		}
		return folder
	}

	fun file(parent: WebDavFolder, name: String, size: Long?): WebDavFile {
		return WebDavFile(parent, name, parent.path + '/' + name, size, null)
	}

	fun folder(parent: WebDavFolder, name: String): WebDavFolder {
		return WebDavFolder(parent, name, parent.path + '/' + name)
	}

	@Throws(BackendException::class)
	fun exists(node: WebDavNode): Boolean {
		node.parent?.let {
			return try {
				connectionHandler.get(absoluteUriFrom(node.path), it) != null
			} catch (e: NotFoundException) {
				false
			}
		} ?: throw ParentFolderIsNullException(node.name)
	}

	@Throws(BackendException::class)
	fun list(folder: WebDavFolder): List<WebDavNode> {
		return connectionHandler.dirList(absoluteUriFrom(folder.path), folder)
	}

	@Throws(BackendException::class)
	fun create(folder: WebDavFolder): WebDavFolder {
		return try {
			createExcludingParents(folder)
		} catch (e: NotFoundException) {
			folder.parent?.let {
				create(it)
				createExcludingParents(folder)
			} ?: throw ParentFolderIsNullException(folder.name)
		} catch (e: ParentFolderDoesNotExistException) {
			folder.parent?.let {
				create(it)
				createExcludingParents(folder)
			} ?: throw ParentFolderIsNullException(folder.name)
		}
	}

	@Throws(BackendException::class)
	private fun createExcludingParents(folder: WebDavFolder): WebDavFolder {
		return if (folder.parent == null) {
			folder
		} else {
			connectionHandler.createFolder(absoluteUriFrom(folder.path), folder)
		}
	}

	@Throws(BackendException::class)
	fun move(source: WebDavFolder, target: WebDavFolder): WebDavFolder {
		moveFileOrFolder(source, target)
		return WebDavFolder(
			target.parent, target.name, target.path
		)
	}

	@Throws(BackendException::class)
	fun move(source: WebDavFile, target: WebDavFile): WebDavFile {
		moveFileOrFolder(source, target)
		return WebDavFile(target.parent, target.name, target.path, source.size, source.modified)
	}

	@Throws(BackendException::class)
	private fun moveFileOrFolder(source: WebDavNode, target: WebDavNode) {
		if (exists(target)) {
			throw CloudNodeAlreadyExistsException(target.name)
		}
		connectionHandler.move(absoluteUriFrom(source.path), absoluteUriFrom(target.path))
	}

	@Throws(BackendException::class, IOException::class)
	fun write(uploadFile: WebDavFile, data: DataSource, progressAware: ProgressAware<UploadState>, replace: Boolean, size: Long): WebDavFile {
		if (!replace && exists(uploadFile)) {
			throw CloudNodeAlreadyExistsException("CloudNode already exists and replace is false")
		}

		progressAware.onProgress(Progress.started(UploadState.upload(uploadFile)))
		data.open(context)?.use { inputStream ->
			object : TransferredBytesAwareInputStream(inputStream) {
				override fun bytesTransferred(transferred: Long) {
					progressAware.onProgress( //
						Progress.progress(UploadState.upload(uploadFile)) //
							.between(0) //
							.and(size) //
							.withValue(transferred)
					)
				}
			}.use {
				connectionHandler.writeFile(absoluteUriFrom(uploadFile.path), it)
			}
		} ?: throw FatalBackendException("InputStream shouldn't bee null")

		return connectionHandler.get(absoluteUriFrom(uploadFile.path), uploadFile.parent) as WebDavFile? ?: throw FatalBackendException("Unable to get CloudFile after upload.")
	}

	@Throws(BackendException::class)
	fun checkAuthenticationAndServerCompatibility(url: String) {
		try {
			connectionHandler.checkAuthenticationAndServerCompatibility(url)
		} catch (ex: ServerNotWebdavCompatibleException) {
			throw WebDavNotSupportedException(cloud)
		}
	}

	@Throws(BackendException::class, IOException::class)
	fun read(file: CloudFile, data: OutputStream, progressAware: ProgressAware<DownloadState>) {
		progressAware.onProgress(Progress.started(DownloadState.download(file)))
		connectionHandler.readFile(absoluteUriFrom(file.path)).use { inputStream ->
			object : TransferredBytesAwareOutputStream(data) {
				override fun bytesTransferred(transferred: Long) {
					progressAware.onProgress( //
						Progress.progress(DownloadState.download(file)) //
							.between(0) //
							.and(file.size ?: Long.MAX_VALUE) //
							.withValue(transferred)
					)
				}
			}.use { out -> CopyStream.copyStreamToStream(inputStream, out) }
		}
		progressAware.onProgress(Progress.completed(DownloadState.download(file)))
	}

	@Throws(BackendException::class)
	fun delete(node: CloudNode) {
		connectionHandler.delete(absoluteUriFrom(node.path))
	}

	private fun absoluteUriFrom(path: String): String {
		return baseUrl //
			.newBuilder() //
			.addPathSegments(path.removePrefix("/")) //
			.build() //
			.toString()
	}

	@Throws(BackendException::class)
	fun currentAccount(): String {
		checkAuthenticationAndServerCompatibility(cloud.url())
		return cloud.url()
	}

}
